components/node-labeler/cmd/run.go

```diff
@@ -197,90 +224,183 @@ func (r *PodReconciler) Reconcile(ctx context.Context, req reconcile.Request) (r
+		return true, nil
 	}
 
 	if !pod.ObjectMeta.DeletionTimestamp.IsZero() {
 		// the pod is being removed.
-		// remove the component label from the node
-		time.Sleep(1 * time.Second)
-		err := updateLabel(labelToUpdate, false, nodeName, r)
-		if err != nil {
-			// this is a edge case when cluster-autoscaler removes a node
-			// (all the running pods will be removed after that)
-			if errors.IsNotFound(err) {
-				return reconcile.Result{}, nil
-			}
-
-			log.WithError(err).Error("removing node label")
-			return reconcile.Result{RequeueAfter: defaultRequeueTime}, err
-		}
-
-		return reconcile.Result{}, err
+		// add the taint to the node
+		return false, nil
 	}
 
 	if !IsPodReady(pod) {
 		// not ready. Wait until the next update.
-		return reconcile.Result{}, nil
-	}
-
-	var node corev1.Node
-	err = r.Get(ctx, types.NamespacedName{Name: nodeName}, &node)
-	if err != nil {
-		return reconcile.Result{}, fmt.Errorf("obtaining node %s: %w", nodeName, err)
-	}
-
-	if labelValue, exists := node.Labels[labelToUpdate]; exists && labelValue == "true" {
-		// nothing to do, the label already exists.
-		return reconcile.Result{}, nil
+		return false, nil
 	}
 
-	err = checkTCPPortIsReachable(ipAddress, port)
+	err := checkTCPPortIsReachable(ipAddress, port)
 	if err != nil {
 		log.WithField("host", ipAddress).WithField("port", port).WithField("pod", pod.Name).WithError(err).Error("checking if TCP port is open")
-		return reconcile.Result{RequeueAfter: defaultRequeueTime}, nil
+		return false, nil
 	}
 
-	if component == registryFacade {
+	if strings.HasPrefix(pod.Name, registryFacade) {
 		err = checkRegistryFacade(ipAddress, port)
 		if err != nil {
 			log.WithError(err).Error("checking registry-facade")
-			return reconcile.Result{RequeueAfter: defaultRequeueTime}, nil
+			return false, nil
 		}
+	}
+
+	return true, nil
+}
 
-		time.Sleep(1 * time.Second)
+type NodeReconciler struct {
+	client.Client
+}
+
+func (r *NodeReconciler) nodeFilter() predicate.Predicate {
+	return predicate.Funcs{
+		CreateFunc: func(e event.CreateEvent) bool {
+			node, ok := e.Object.(*corev1.Node)
+			if !ok {
+				return false
+			}
+			return isWorkspaceNode(*node)
+		},
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			return false
+		},
+		DeleteFunc: func(e event.DeleteEvent) bool {
+			return false
+		},
+	}
+}
+
+func (r *NodeReconciler) reconcileAll(ctx context.Context) error {
+	log.Info("start reconciling all nodes")
+
+	var nodes corev1.NodeList
+	if err := r.List(ctx, &nodes); err != nil {
+		return fmt.Errorf("failed to list nodes: %w", err)
 	}
 
-	err = updateLabel(labelToUpdate, true, nodeName, r)
-	if err != nil {
-		log.WithError(err).Error("updating node label")
-		return reconcile.Result{}, fmt.Errorf("trying to add the label: %v", err)
+	for _, node := range nodes.Items {
+		if node.Labels == nil {
+			continue
+		}
+		if !isWorkspaceNode(node) {
+			continue
+		}
+
+		err := updateNodeLabel(node.Name, r.Client)
```

Comment:
Why are we updating labels on the nodes? I thought taint would replace.

Reply:
How do we test this flow that you are describing? It seems possible, but low percentage.

Reply:
> The possible side effect is that an additional node may be started but shut down after five minutes. 

OK, this happens now, so not really an issue, I  think. Let's do in follow-on.